testing/test_kpis.py

```diff
@@ -93,15 +93,15 @@ def test_get_cost(self):
         # Check results
         self._perform_test(self.case.cost_tot, self.case.cost_dict, 'cost_constant')
 
-        # Reset KPI Calculator
-        self.cal.initialize()
+        # Reset testcase and kpi calculator
+        self._initialize_testcase(self.testcase_name, self.ref_data_filepath)
```

Comment:
I'm struggling a bit with this one. Is it really needed to initialize the testcase as well? actually, is it correct? does not the simulation data get lost when re-initializing the case? Apparently not since tests are passing, but that's surprising to me.

Reply:
I see now, simulation data is not lost because it is reloaded in `_initialize_testcase()`. Still, wouldn't we get the same result by just resetting the KPI Caluclator with `self.cal.initialize()`? I like more that approach since that also checks that the KPI Calculator can re-calculate KPIs without having to re-run the simulation even after changing the scenario.

Reply:
No you do not get the same result.  Using ``self.cal.initialize()`` resets the stored kpi value to 0 using ``initialize_kpi_vars`` and resets the last index using ``_set_last_index``.  The last index will point to the end of the already existing ``y_store``.  So, when the kpi calculator is invoked the next time, you will get the kpi to be equal to 0 + 0 = 0.  The first 0 being from the reset of the kpi value itself, and the second 0 being from the fact that the the last index is already at the end of the ``y_store``. 

So, in the current implementation, you cannot change the scenario and re-calculate the KPIs from the beginning of the simulation (while also excluding the initialization period) with the new scenario.  To do so, we'd have to implement a way for ``_set_last_index`` to save the index it obtained when the test case itself was initialized, and then know to go back to that index if only the kpi calculator is re-initialized (and not also the test case).  I guess the question is how needed is that functionality.  I guess the one use case I think of is if you want to re-calculate the cost KPI for an alternative scenario then the one you ran your controller for (which may or may not have used the initial price scenario for control).

Reply:
I see now, thanks for the explanation! I need to sleep on it a bit. Let me come back to it next Monday.